python/private/bzlmod/pypi_archive.bzl

```diff
@@ -0,0 +1,137 @@
+    prefix, _, _ = prefix.rpartition("_")
+
+    _, _, filename = rctx.attr.urls[0].rpartition("/")
+    filename = filename.strip()
+
+    urls = rctx.attr.urls
+    auth = get_auth(rctx, urls)
+
+    result = rctx.download(
+        url = urls,
+        output = filename,
+        sha256 = rctx.attr.sha256,
+        auth = auth,
+        canonical_id = rctx.attr.canonical_id,
+        integrity = rctx.attr.integrity,
+    )
+    if not result.success:
+        fail(result)
+
+    whl_path = rctx.path(filename)
+
+    if rctx.attr.patches:
+        whl_path = patch_whl(
+            rctx,
+            python_interpreter = _resolve_python_interpreter(rctx),
+            whl_path = whl_path,
+            patches = rctx.attr.patches,
+            quiet = rctx.attr.quiet,
+        )
+
+    rctx.symlink(whl_path, "file")
+
+    rctx.file(
+        "BUILD.bazel",
+        """\
+filegroup(
+    name="file",
+    srcs=["{filename}"],
+    visibility=["//visibility:public"],
+)
+""".format(filename = whl_path.basename),
+    )
+
+_pypi_file_attrs = {
+    "auth_patterns": attr.string_dict(doc = _HTTP_FILE_DOC),
+    "canonical_id": attr.string(doc = _HTTP_FILE_DOC),
+    "integrity": attr.string(doc = _HTTP_FILE_DOC),
+    "netrc": attr.string(doc = _HTTP_FILE_DOC),
+    "patches": attr.label_keyed_string_dict(
+        doc = """\
+A label-keyed-string dict that has patch_strip as the value and the patch to be applied as
+a label. The patches are applied in the same order as they are listed in the dictionary.
+""",
+    ),
+    "python_interpreter": attr.string(doc = "The python interpreter to use when patching"),
+    "python_interpreter_target": attr.label(doc = "The python interpreter target to use when patching"),
+    "quiet": attr.bool(doc = "Silence the stdout/stdeer during patching", default = True),
+    "sha256": attr.string(doc = _HTTP_FILE_DOC),
+    "urls": attr.string_list(doc = _HTTP_FILE_DOC),
+}
+
+pypi_file = repository_rule(
+    attrs = _pypi_file_attrs,
+    doc = """A rule for downloading a single file from a PyPI like index.""",
+    implementation = _impl,
+)
+
+# TODO @aignas 2023-12-16: expose getting interpreter
+def _get_python_interpreter_attr(rctx):
+    """A helper function for getting the `python_interpreter` attribute or it's default
+
+    Args:
+        rctx (repository_ctx): Handle to the rule repository context.
+
+    Returns:
+        str: The attribute value or it's default
+    """
+    if rctx.attr.python_interpreter:
+        return rctx.attr.python_interpreter
+
+    if "win" in rctx.os.name:
+        return "python.exe"
+    else:
+        return "python3"
+
+def _resolve_python_interpreter(rctx):
+    """Helper function to find the python interpreter from the common attributes
+
+    Args:
+        rctx: Handle to the rule repository context.
+    Returns: Python interpreter path.
+    """
+    python_interpreter = _get_python_interpreter_attr(rctx)
+
+    if rctx.attr.python_interpreter_target != None:
```

Comment:
Also check attr.python_interpreter -- it could be provided as a target or local path.

That said, a lot of this could go away if we just pick a single interpreter to use for our internal tools and hard code that.